src/test/java/org/jetbrains/kotlinx/lincheck/test/verifier/nesting_safe_recoverable_linearizability/set/NRLSet.java

```diff
@@ -0,0 +1,234 @@
+package org.jetbrains.kotlinx.lincheck.test.verifier.nesting_safe_recoverable_linearizability.set;
+
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicMarkableReference;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * @see  <a href="https://arxiv.org/pdf/1905.13600.pdf">Tracking in Order to Recover: Recoverable Lock-Free Data Structures</a>
+ */
+public class NRLSet<T extends Comparable<T>> {
+
+    private class Info {
+        @Nullable
+        public Node node;
+
+        @Nullable
+        public Boolean result;
+
+        public Info() {
+        }
+
+        public Info(@Nullable Node node) {
+            this.node = node;
+        }
+    }
+
+    private class Node {
+        public static final int NULL_DELETER = -1;
+
+        @NotNull
+        final public T value;
+        @NotNull
+        final public AtomicMarkableReference<Node> next;
+        @NotNull
+        final public AtomicInteger deleter = new AtomicInteger(NULL_DELETER);
```

Comment:
Let's use `kotlinx.atomicfu` primitives instead -- they do not produce additional objects